### PR TITLE
add Render to Deployment Guides; close #3683

### LINKED ISSUE #3683: Create a Deployment Guide for Render.com
I'd love to see a Deployment Guide for Hasura on Render.com

### PR DESCRIPTION
**Pull request by ericmand**
#3683 

### Description
Provides a Deployment Guide for deploying Hasura to Render.com

### Affected components
- [X] Docs

### Related Issues
#3683 

#### Catalog upgrade
Does this PR change Hasura Catalog version?
- [X] No

Does this PR add a new Metadata feature?
- [X] No

#### Breaking changes
- [X] No Breaking changes

### CODE REVIEW

**Comment by ericmand** [comment]
I have screenshots that could be included in the Guide but I wasn't sure how to add those.

**Comment by ericmand** [comment]
This Guide would be even easier if we referenced a repo to fork like this one: https://github.com/ericmand/Hasura-on-Render

File: docs/graphql/manual/guides/deployment/render.rst [hunk 1]
```diff
@@ -0,0 +1,115 @@
+Hasura GraphQL Engine on Render
```
Comment (marionschleifer) [inline docs/graphql/manual/guides/deployment/render.rst]:
```suggestion
Deploying Hasura GraphQL engine on Render
```

File: docs/graphql/manual/guides/deployment/render.rst [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+====================================================================
```
Comment (marionschleifer) [inline docs/graphql/manual/guides/deployment/render.rst]: The underlines should have the same length as the titles.

File: docs/graphql/manual/guides/deployment/index.rst [hunk 3]
```diff
@@ -13,6 +13,7 @@ Guides: Deployment
 - :doc:`Digital Ocean One-click App on Marketplace <digital-ocean-one-click>`
 - :doc:`Azure Container Instances with Postgres <azure-container-instances-postgres>`
 - :doc:`Google Cloud Platform with Kubernetes engine and Cloud SQL <google-kubernetes-engine-cloud-sql>`
+- :doc:`Render <render>`
```
Comment (marionschleifer) [inline docs/graphql/manual/guides/deployment/index.rst]: Can we have the full title here, like the others.

File: docs/graphql/manual/guides/deployment/render.rst [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+  :local:
+
+This guide talks about how to deploy the Hasura GraphQL engine on `Render
+<https://render.com>`__.
+
+
+Pre-requisites
+--------------
+
+- GitHub (or alternative) account
```
Comment (marionschleifer) [inline docs/graphql/manual/guides/deployment/render.rst]: What is meant by "alternative"?

File: docs/graphql/manual/guides/deployment/render.rst [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+- Render account associated with a Github account and with billing enabled
+
+
+Create a repository
+-------------------
+
+Create a new GitHub repository
```
Comment (marionschleifer) [inline docs/graphql/manual/guides/deployment/render.rst]: Can we have bullet points on the instructions here.

File: docs/graphql/manual/guides/deployment/render.rst [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+Within the new repository, create a new file named ``Dockerfile`` that 
+references the latest Hasura GraphQL Engine docker image by including the
```
Comment (marionschleifer) [inline docs/graphql/manual/guides/deployment/render.rst]:
```suggestion
references the latest Hasura GraphQL engine Docker image by including the 
```

File: docs/graphql/manual/guides/deployment/render.rst [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+Within the new repository, create a new file named ``Dockerfile`` that 
+references the latest Hasura GraphQL Engine docker image by including the 
+following single line in the file:
+
+.. code-block:: bash
+
+   FROM hasura/graphql-engine:latest
+
+
+Using Render Infrastrcuture as Code (Fastest)
```
Comment (marionschleifer) [inline docs/graphql/manual/guides/deployment/render.rst]: Not sure what is meant by "using Render infrastructure as code"

File: docs/graphql/manual/guides/deployment/render.rst [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+---------------------------------------------
+
+In the same repository, create a second new file, this one named
```
Comment (marionschleifer) [inline docs/graphql/manual/guides/deployment/render.rst]:
```suggestion
In the same repository, create a second new file named 
```

File: docs/graphql/manual/guides/deployment/render.rst [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+.. code-block:: bash
+
+   FROM hasura/graphql-engine:latest
+
+
+Using Render Infrastrcuture as Code (Fastest)
+---------------------------------------------
+
+In the same repository, create a second new file, this one named 
+``render.yaml``. Includes the following Render Infrastrcuture as
```
Comment (marionschleifer) [inline docs/graphql/manual/guides/deployment/render.rst]:
```suggestion
``render.yaml``. Insert the following Render infrastructure as 
```

File: docs/graphql/manual/guides/deployment/render.rst [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+
+   FROM hasura/graphql-engine:latest
+
+
+Using Render Infrastrcuture as Code (Fastest)
+---------------------------------------------
+
+In the same repository, create a second new file, this one named 
+``render.yaml``. Includes the following Render Infrastrcuture as 
+Code configuration:
```
Comment (marionschleifer) [inline docs/graphql/manual/guides/deployment/render.rst]:
```suggestion
code configuration:
```

File: docs/graphql/manual/guides/deployment/render.rst [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+     - key: HASURA_GRAPHQL_DATABASE_URL
+       fromDatabase:
+         name: hasura
+         property: connectionString
+     - key: HASURA_GRAPHQL_ENABLE_CONSOLE
+       value: true
+   databases:
+   - name: hasura
+
+Within the Render Dashboard, find the YAML section and the New from
```
Comment (marionschleifer) [inline docs/graphql/manual/guides/deployment/render.rst]:
```suggestion
On the Render dashboard, find the **YAML** section and the **New from 
```

File: docs/graphql/manual/guides/deployment/render.rst [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+       fromDatabase:
+         name: hasura
+         property: connectionString
+     - key: HASURA_GRAPHQL_ENABLE_CONSOLE
+       value: true
+   databases:
+   - name: hasura
+
+Within the Render Dashboard, find the YAML section and the New from 
+YAML button or `go directly there <https://dashboard.render.com/select-repo?type=iac>`_
```
Comment (marionschleifer) [inline docs/graphql/manual/guides/deployment/render.rst]:
```suggestion
YAML** button or `go directly there <https://dashboard.render.com/select-repo?type=iac>`_ 
```

File: docs/graphql/manual/guides/deployment/render.rst [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+     - key: HASURA_GRAPHQL_ENABLE_CONSOLE
+       value: true
+   databases:
+   - name: hasura
+
+Within the Render Dashboard, find the YAML section and the New from 
+YAML button or `go directly there <https://dashboard.render.com/select-repo?type=iac>`_ 
+and select the repository you've created.
+
+This it! Skip to notes about the Hasura Console and debugging with logs below.
```
Comment (marionschleifer) [inline docs/graphql/manual/guides/deployment/render.rst]:
```suggestion
That's it! Skip to notes about the Hasura console and debug with the logs below.
```

Comment (marionschleifer) [inline docs/graphql/manual/guides/deployment/render.rst]: What is meant by "skip to notes about the Hasura console"?

File: docs/graphql/manual/guides/deployment/render.rst [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+
+
+Provision the Database and Web Service Manually (Alternate Method)
```
Comment (marionschleifer) [inline docs/graphql/manual/guides/deployment/render.rst]:
```suggestion
Provision the database and web service manually (alternative method)
```

File: docs/graphql/manual/guides/deployment/render.rst [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+------------------------------------------------------------------
+
+.. note::
+
+   If you already have a Render database setup, you can skip this step.
+
+Create a Postgres server by providing a name in Render's `New Database
```
Comment (marionschleifer) [inline docs/graphql/manual/guides/deployment/render.rst]: Can we have bullet points for the instructions here?

File: docs/graphql/manual/guides/deployment/render.rst [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+
+This it! Skip to notes about the Hasura Console and debugging with logs below.
+
+
+Provision the Database and Web Service Manually (Alternate Method)
+------------------------------------------------------------------
+
+.. note::
+
+   If you already have a Render database setup, you can skip this step.
```
Comment (marionschleifer) [inline docs/graphql/manual/guides/deployment/render.rst]:
```suggestion
   If you already have a Render database setup, you can skip the first step.
```

File: docs/graphql/manual/guides/deployment/render.rst [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+
+Create a Postgres server by providing a name in Render's `New Database 
+configuration page <https://dashboard.render.com/new/database>`__.
+
+Once created, you'll see an Internal Connection String which
```
Comment (marionschleifer) [inline docs/graphql/manual/guides/deployment/render.rst]:
```suggestion
Once created, you'll see an internal connection string which 
```

File: docs/graphql/manual/guides/deployment/render.rst [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+configuration page <https://dashboard.render.com/new/database>`__.
+
+Once created, you'll see an Internal Connection String which 
+is the database URL used in the next step.
+
+Create a web service by selecting the repository you just created in the `New Web 
+Service configuration page <https://dashboard.render.com/select-repo?type=web>`__. 
+You have to specify a name.
+
+Upon creation of the web service, find the Environment tab and create two
```
Comment (marionschleifer) [inline docs/graphql/manual/guides/deployment/render.rst]:
```suggestion
Upon creation of the web service, find the environment tab and create two 
```

File: docs/graphql/manual/guides/deployment/render.rst [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+Once created, you'll see an Internal Connection String which 
+is the database URL used in the next step.
+
+Create a web service by selecting the repository you just created in the `New Web 
+Service configuration page <https://dashboard.render.com/select-repo?type=web>`__. 
+You have to specify a name.
+
+Upon creation of the web service, find the Environment tab and create two 
+environment variables: one named ``HASURA_GRAPHQL_DATABASE_URL`` whose value is 
+the Internal Connection String from the database you created earlier;
```
Comment (marionschleifer) [inline docs/graphql/manual/guides/deployment/render.rst]:
```suggestion
the internal connection string from the database you created earlier, 
```

File: docs/graphql/manual/guides/deployment/render.rst [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+is the database URL used in the next step.
+
+Create a web service by selecting the repository you just created in the `New Web 
+Service configuration page <https://dashboard.render.com/select-repo?type=web>`__. 
+You have to specify a name.
+
+Upon creation of the web service, find the Environment tab and create two 
+environment variables: one named ``HASURA_GRAPHQL_DATABASE_URL`` whose value is 
+the Internal Connection String from the database you created earlier; 
+another named ``HASURA_GRAPHQL_ENABLE_CONSOLE`` whose value is ``true``.
```
Comment (marionschleifer) [inline docs/graphql/manual/guides/deployment/render.rst]:
```suggestion
and another named ``HASURA_GRAPHQL_ENABLE_CONSOLE`` whose value is ``true``.
```

File: docs/graphql/manual/guides/deployment/render.rst [hunk 21]
```diff
@@ -0,0 +1,115 @@
...
+
+
+Open the Hasura Console
+-----------------------
+
+That's it! You can monitor the deployment of the web service from the Logs tab.
```
Comment (marionschleifer) [inline docs/graphql/manual/guides/deployment/render.rst]:
```suggestion
That's it! You can monitor the deployment of the web service from the logs tab. 
```

File: docs/graphql/manual/guides/deployment/render.rst [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+environment variables: one named ``HASURA_GRAPHQL_DATABASE_URL`` whose value is 
+the Internal Connection String from the database you created earlier; 
+another named ``HASURA_GRAPHQL_ENABLE_CONSOLE`` whose value is ``true``.
+
+
+Open the Hasura Console
+-----------------------
+
+That's it! You can monitor the deployment of the web service from the Logs tab. 
+There you can see error messages if deployment fails. If deployment succeeds, you
```
Comment (marionschleifer) [inline docs/graphql/manual/guides/deployment/render.rst]:
```suggestion
There you can see error messages if the deployment fails. If the deployment succeeds, you 
```

File: docs/graphql/manual/guides/deployment/render.rst [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+URL for the Hasura console:
+
+.. code:: 
+
+   http://<web-service-name>.onrender.com
+
+Replace ``<web-service-name>`` with the name specified earlier.
+
+You can create tables and test your GraphQL queries here. Check out :ref:`Making
+your first GraphQL Query <first_graphql_query>` for a detailed guide.
```
Comment (marionschleifer) [inline docs/graphql/manual/guides/deployment/render.rst]:
```suggestion
your first GraphQL Query <first_graphql_query>` for a detailed example.
```

File: docs/graphql/manual/guides/deployment/render.rst [hunk 24]
```diff
@@ -0,0 +1,115 @@
...
+
+
+Tearing down
+------------
+
+To clean-up, delete the web service and the database from the Render Dashboard.
```
Comment (marionschleifer) [inline docs/graphql/manual/guides/deployment/render.rst]:
```suggestion
To clean up, delete the web service and the database from the Render dashboard.
```

File: docs/graphql/manual/guides/deployment/render.rst (hunk 4 above)
Comment (ericmand) [reply docs/graphql/manual/guides/deployment/render.rst]: Gitlab, mostly, but possibly Bitbucket as well. I didn't take the time to confirm it is the same process. We could leave the bit about alternates out.

File: docs/graphql/manual/guides/deployment/render.rst (hunk 7 above)
Comment (ericmand) [reply docs/graphql/manual/guides/deployment/render.rst]: This is a capabiltiy of Render. I'm not sure if we should refer to it as a product name "Infrastructure as Code" (capitalized proper noun) or just a feature description (lowercase)

File: docs/graphql/manual/guides/deployment/render.rst (hunk 13 above)
Comment (ericmand) [reply docs/graphql/manual/guides/deployment/render.rst]: Skip to the next section titled "Provision the Database and Web Service Manually (Alternate Method)" and proceed to the following section titled "Open the Hasura Console"

File: docs/graphql/manual/guides/deployment/index.rst (hunk 3 above)
Comment (ericmand) [reply docs/graphql/manual/guides/deployment/index.rst]: Yes, except I'm not sure what I would add to it. I suppose we could say "Render with Postgres" or "Render Web Services with Postgres" but the Render product suite is simple enough that the qualification may not be necessary. There is only one way to do this on Render. @anurag May have suggestions here.

Comment (anurag) [reply docs/graphql/manual/guides/deployment/index.rst]: How about `Render Single Click Deploy with PostgreSQL`

File: docs/graphql/manual/guides/deployment/render.rst (hunk 4 above)
Comment (anurag) [reply docs/graphql/manual/guides/deployment/render.rst]: `GitHub` or `GitLab` account.

File: docs/graphql/manual/guides/deployment/render.rst (hunk 7 above)
Comment (anurag) [reply docs/graphql/manual/guides/deployment/render.rst]: I think we can change this to Using `render.yaml` (fastest)

File: docs/graphql/manual/guides/deployment/render.rst (hunk 5 above)
Comment (ericmand) [reply docs/graphql/manual/guides/deployment/render.rst]: Attempted fix, but I'm worried I didn't do this right.

**Comment by anurag** [comment]
@marionschleifer I can take point on this PR going forward. Thanks for submitting it @ericmand !

**Comment by marionschleifer** [comment]
@ericmand @anurag I just pushed to the branch fixing:

- bullet points
- broken links
- render guide showing up in the navigation (see below)

image

Otherwise, this looks good. However, as you mentioned, it would be cool to have some screenshots. 

If you can share the screenshots with me and tell me where to put those, I'm happy to add them. 2-3 will be enough 🙂

**Comment by anurag** [comment]
Thanks @marionschleifer. I can post the screenshots. However, we recently released a `Deploy To Render` button which will make it even easier for Hasura users to host it on Render. They also won't need to fork or create a repo. I should have the code/repo with the deploy to render button ready to go later this week. Does it make sense to open that PR instead?

**Comment by anurag** [comment]
You can see the button in action here: https://github.com/tryretool/retool-onpremise#deploying-to-render

**Comment by marionschleifer** [comment]
@ericmand what do you mean by `Does it make sense to open that PR instead?`? I think we can just wait until you have things ready at the end of the week and then modify this PR? Does that work?

**Comment by anurag** [comment]
> @ericmand what do you mean by `Does it make sense to open that PR instead?`? I think we can just wait until you have things ready at the end of the week and then modify this PR? Does that work?

Sure. The problem is that I (@anurag) don't have push access to Eric's fork, and the new PR is going to be very different in terms of content vs. this PR. So I was planning to create my own branch/fork and create a new PR with my changes. I've signed the CLA. We can close this PR when the new one is up. Does that work?

**Comment by marionschleifer** [comment]
@anurag ah okay, got it. Yes, that makes sense 🙂

**Comment by anurag** [comment]
@marionschleifer the new PR is available at https://github.com/hasura/graphql-engine/pull/4209

**Comment by marionschleifer** [comment]
@anurag thank you. I'm closing this PR in favour of #4209.

**Pull request closed without merging**